File: sample_question/custom_validator/sample-validator.hs
import Control.Monad (when)
import Data.Char (isDigit, isSpace)
import Data.Maybe (fromMaybe)
import System.Environment (lookupEnv)
import System.Exit (ExitCode (..), exitWith)
import System.IO

acCode, waCode, tleCode, reCode, peCode :: Int
acCode = 0
waCode = 1
tleCode = 2
reCode = 3
peCode = 4

finish :: Int -> Maybe String -> IO a
finish exitcode message = do
  case message of
    Just text -> hPutStrLn stderr text >> hFlush stderr
    Nothing -> return ()
  if exitcode == 0 then exitWith ExitSuccess else exitWith (ExitFailure exitcode)

readLineMaybe :: IO (Maybe String)
readLineMaybe = do
  eof <- isEOF
  if eof then return Nothing else Just . trim <$> getLine

trim :: String -> String
trim = dropWhile isSpace . reverse . dropWhile isSpace . reverse

parseInteger :: String -> Maybe Integer
parseInteger text =
  case trim text of
    ('-' : rest) | not (null rest) && all isDigit rest -> Just (negate (read rest))
    ('+' : rest) | not (null rest) && all isDigit rest -> Just (read rest)
    digits | not (null digits) && all isDigit digits -> Just (read digits)
    _ -> Nothing

envNumber :: String -> IO String
envNumber name = fromMaybe "null" <$> lookupEnv name

printLimits :: IO ()
printLimits = do
  timeLimit <- envNumber "PROBLEM_TIME_LIMIT"
  outputLimit <- envNumber "PROBLEM_OUTPUT_LIMIT"
  memoryLimit <- envNumber "PROBLEM_MEMORY_LIMIT"
  pidLimit <- envNumber "PROBLEM_PID_LIMIT"
  userLanguage <- maybe "null" (\value -> "\"" ++ value ++ "\"") <$> lookupEnv "USER_LANGUAGE"
  perLanguageLimits <- fromMaybe "null" <$> lookupEnv "PER_LANGUAGE_LIMITS"
  hPutStrLn stderr $
    "{\"problem_time_limit\": " ++ timeLimit
      ++ ", \"problem_output_limit\": " ++ outputLimit
      ++ ", \"problem_memory_limit\": " ++ memoryLimit
      ++ ", \"problem_pid_limit\": " ++ pidLimit
      ++ ", \"user_language\": " ++ userLanguage
      ++ ", \"per_language_limits\": " ++ perLanguageLimits
      ++ "}"
  hFlush stderr

readSecretValue :: IO Integer
readSecretValue = do
  line <- readLineMaybe
  case line >>= parseInteger of
    Just value -> return value
    Nothing -> finish reCode (Just "Invalid secret data")

validate :: Integer -> Integer -> Integer -> IO ()
validate maxTries secret iteration = do
  line <- readLineMaybe
  case line of
    Nothing -> finish reCode (Just "Unexpected flow")
    Just user -> do
      case user of
        ('!' : answerText) ->
          case parseInteger answerText of
            Nothing -> finish reCode (Just "Invalid data")
            Just answer ->
              if answer == secret
                then finish acCode Nothing
                else finish waCode (Just ("!" ++ show secret))
        _ -> return ()

      when (iteration > maxTries) $
        finish tleCode (Just "Exceeded number of allowed iterations")

      case parseInteger user of
        Nothing -> finish reCode (Just "Invalid data")
        Just value -> do
          putStrLn (if secret < value then "<" else ">=")
          hFlush stdout
          validate maxTries secret (iteration + 1)

main :: IO ()
main = do
  hSetBuffering stdout LineBuffering
  maxValue <- readSecretValue
  maxTries <- readSecretValue
  secret <- readSecretValue

  printLimits

  print maxValue
  hFlush stdout

  validate maxTries secret 0

File: sample_question/custom_validator/sample-validator.pl
:- initialization(main, main).
:- use_module(library(readutil)).

exit_code(ac, 0).
exit_code(wa, 1).
exit_code(tle, 2).
exit_code(re, 3).
exit_code(pe, 4).

finish(Verdict) :-
    exit_code(Verdict, Code),
    halt(Code).

finish(Verdict, Message) :-
    format(user_error, "~w~n", [Message]),
    flush_output(user_error),
    exit_code(Verdict, Code),
    halt(Code).

read_trimmed_line(Line) :-
    read_line_to_string(current_input, Raw),
    (   Raw == end_of_file
    ->  Line = end_of_file
    ;   normalize_space(string(Line), Raw)
    ).

parse_integer(Text, Value) :-
    catch(number_string(Value, Text), _, fail),
    integer(Value).

env_number(Name, Value) :-
    (   getenv(Name, Raw)
    ->  Value = Raw
    ;   Value = 'null'
    ).

print_limits :-
    env_number('PROBLEM_TIME_LIMIT', TimeLimit),
    env_number('PROBLEM_OUTPUT_LIMIT', OutputLimit),
    env_number('PROBLEM_MEMORY_LIMIT', MemoryLimit),
    env_number('PROBLEM_PID_LIMIT', PidLimit),
    (   getenv('USER_LANGUAGE', Language)
    ->  format(atom(UserLanguage), "\"~w\"", [Language])
    ;   UserLanguage = 'null'
    ),
    env_number('PER_LANGUAGE_LIMITS', PerLanguageLimits),
    format(user_error,
           '{"problem_time_limit": ~w, "problem_output_limit": ~w, "problem_memory_limit": ~w, \c
            "problem_pid_limit": ~w, "user_language": ~w, "per_language_limits": ~w}~n',
           [TimeLimit, OutputLimit, MemoryLimit, PidLimit, UserLanguage, PerLanguageLimits]),
    flush_output(user_error).

read_secret_value(Value) :-
    read_trimmed_line(Line),
    (   Line \== end_of_file,
        parse_integer(Line, Parsed)
    ->  Value = Parsed
    ;   finish(re, 'Invalid secret data')
    ).

main :-
    read_secret_value(MaxValue),
    read_secret_value(MaxTries),
    read_secret_value(Secret),
    print_limits,
    format("~d~n", [MaxValue]),
    flush_output,
    validate(MaxTries, Secret, 0).

validate(MaxTries, Secret, Iteration) :-
    read_trimmed_line(Line),
    (   Line == end_of_file
    ->  finish(re, 'Unexpected flow')
    ;   handle_answer(Line, Secret),
        check_budget(Iteration, MaxTries),
        respond(Line, Secret),
        Next is Iteration + 1,
        validate(MaxTries, Secret, Next)
    ).

handle_answer(Line, Secret) :-
    (   string_concat("!", AnswerText, Line)
    ->  (   parse_integer(AnswerText, Answer)
        ->  (   Answer =:= Secret
            ->  finish(ac)
            ;   format(atom(Message), "!~d", [Secret]),
                finish(wa, Message)
            )
        ;   finish(re, 'Invalid data')
        )
    ;   true
    ).

check_budget(Iteration, MaxTries) :-
    (   Iteration > MaxTries
    ->  finish(tle, 'Exceeded number of allowed iterations')
    ;   true
    ).

respond(Line, Secret) :-
    (   parse_integer(Line, Value)
    ->  (   Secret < Value
        ->  format("<~n")
        ;   format(">=~n")
        ),
        flush_output
    ;   finish(re, 'Invalid data')
    ).
